HLTrigger/btau/src/HLTDisplacedtktktkFilter.cc

```diff
@@ -0,0 +1,210 @@
+  minCosinePointingAngle_ (iConfig.getParameter<double>("MinCosinePointingAngle")),
+  triggerTypeDaughters_(iConfig.getParameter<int>("triggerTypeDaughters")),
+  DisplacedVertexTag_(iConfig.getParameter<edm::InputTag>("DisplacedVertexTag")),
+  DisplacedVertexToken_(consumes<reco::VertexCollection>(DisplacedVertexTag_)),
+  beamSpotTag_ (iConfig.getParameter<edm::InputTag> ("BeamSpotTag")),
+  beamSpotToken_(consumes<reco::BeamSpot>(beamSpotTag_)),
+  TrackTag_ (iConfig.getParameter<edm::InputTag>("TrackTag")),
+  TrackToken_(consumes<reco::RecoChargedCandidateCollection>(TrackTag_))
+{
+}
+
+
+HLTDisplacedtktktkFilter::~HLTDisplacedtktktkFilter() = default;
+
+void HLTDisplacedtktktkFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  makeHLTFilterDescription(desc);
+  desc.add<bool>("FastAccept",false);
+  desc.add<double>("MinLxySignificance",0.0);
+  desc.add<double>("MaxLxySignificance",0.0);
+  desc.add<double>("MaxNormalisedChi2",10.0);
+  desc.add<double>("MinVtxProbability",0.0);
+  desc.add<double>("MinCosinePointingAngle",-2.0);
+  desc.add<int>("triggerTypeDaughters",0);
+  desc.add<edm::InputTag>("DisplacedVertexTag",edm::InputTag("hltDisplacedtktktkVtx"));
+  desc.add<edm::InputTag>("BeamSpotTag",edm::InputTag("hltOnlineBeamSpot"));
+  desc.add<edm::InputTag>("TrackTag",edm::InputTag("hltL3MuonCandidates"));
+  descriptions.add("hltDisplacedtktktkFilter", desc);
+}
+
+// ------------ method called once each job just before starting event loop  ------------
+void HLTDisplacedtktktkFilter::beginJob()
+{
+
+}
+
+// ------------ method called once each job just after ending the event loop  ------------
+void HLTDisplacedtktktkFilter::endJob()
+{
+ 	
+}
+
+// ------------ method called on each new Event  ------------
+bool HLTDisplacedtktktkFilter::hltFilter(edm::Event& iEvent, const edm::EventSetup& iSetup, trigger::TriggerFilterObjectWithRefs & filterproduct) const
+{
+
+
+  // get beam spot
+  reco::BeamSpot vertexBeamSpot;
+  edm::Handle<reco::BeamSpot> recoBeamSpotHandle;
+  iEvent.getByToken(beamSpotToken_,recoBeamSpotHandle);
+  vertexBeamSpot = *recoBeamSpotHandle;
+
+
+  // get displaced vertices
+  reco::VertexCollection displacedVertexColl;
+  edm::Handle<reco::VertexCollection> displacedVertexCollHandle;
+  bool foundVertexColl = iEvent.getByToken(DisplacedVertexToken_, displacedVertexCollHandle);
+  if(foundVertexColl) displacedVertexColl = *displacedVertexCollHandle;
+
+
+  // get track collection
+  edm::Handle<reco::RecoChargedCandidateCollection> trackcands;
+  iEvent.getByToken(TrackToken_,trackcands);
+
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+
+
+  // Ref to Candidate object to be recorded in filter object
+  reco::RecoChargedCandidateRef ref1;
```

Comment:
ref1,ref2,ref3 can be declared below where they are used

Reply:
changed